### PR TITLE
add support for __builtin_fpclassify

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
 
     return bswap_expr;
   }
+  else if(identifier=="__builtin_fpclassify" ||
+          identifier==CPROVER_PREFIX "fpclassify")
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: Nit pick: clang-format spacing (throughout this PR)

File: src/ansi-c/c_typecheck_expr.cpp [hunk 2]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
...
+      throw 0;
+    }
+
+    const auto &floatbv_type=to_floatbv_type(expr.arguments()[5].type());
+
+    const exprt zero=
+      ieee_floatt::zero(floatbv_type).to_expr();
+
+    return
+      if_exprt(unary_predicate_exprt(ID_isnan, expr.arguments()[5]),
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: `isnan_exprt(expr.arguments()[5])`

File: src/ansi-c/c_typecheck_expr.cpp [hunk 3]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
...
+               expr.arguments()[0],
+      if_exprt(unary_predicate_exprt(ID_isinf, expr.arguments()[5]),
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: `isinf_exprt`

File: src/ansi-c/c_typecheck_expr.cpp [hunk 4]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
...
+               expr.arguments()[1],
+      if_exprt(unary_predicate_exprt(ID_isnormal, expr.arguments()[5]),
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: `isnormal_exprt`

File: src/ansi-c/c_typecheck_expr.cpp [hunk 5]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
 
     return bswap_expr;
   }
+  else if(identifier=="__builtin_fpclassify" ||
+          identifier==CPROVER_PREFIX "fpclassify")
+  {
+    if(expr.arguments().size()!=6)
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: I'd add `const auto &arguments = expr.arguments();` as that function call is repeated a number of times in this function.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 6]
```diff
@@ -2195,6 +2195,44 @@ exprt c_typecheck_baset::do_special_functions(
...
+      error() << identifier << " expects six arguments" << eom;
+      throw 0;
+    }
+
+    // This gets 5 integers followed by a float or double.
+    // The five integers are the return values for the cases
+    // FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL and FP_ZERO.
+    // gcc expects this to be able to produce compile-time constants.
+
+    if(expr.arguments()[5].type().id()!=ID_floatbv)
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: How about `const exprt &input_value = arguments[5];` as that particular argument is used repeatedly?

**Comment by kroening** [comment]
done.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 7]
```diff
@@ -2195,6 +2195,50 @@ exprt c_typecheck_baset::do_special_functions(
...
+      throw 0;
+    }
+
+    // This gets 5 integers followed by a float or double.
+    // The five integers are the return values for the cases
+    // FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL and FP_ZERO.
+    // gcc expects this to be able to produce compile-time constants.
+
+    const auto &arguments = expr.arguments();
+    const exprt &fp_value=arguments[5];
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: Nit pick: spacing

File: src/ansi-c/c_typecheck_expr.cpp [hunk 8]
```diff
@@ -2195,6 +2195,50 @@ exprt c_typecheck_baset::do_special_functions(
...
+      error() << identifier << " expects six arguments" << eom;
+      throw 0;
+    }
+
+    // This gets 5 integers followed by a float or double.
+    // The five integers are the return values for the cases
+    // FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL and FP_ZERO.
+    // gcc expects this to be able to produce compile-time constants.
+
+    const auto &arguments = expr.arguments();
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: I'd personally do this at the very beginning to use it right in the first `if`.

**Review by tautschnig — APPROVED** [review]

**Review by smowton — CHANGES_REQUESTED** [review]
We should make `goto-cc` versions of e.g. https://github.com/diffblue/cbmc/blob/develop/regression/cbmc/Float21/main.c to verify that this works with goto-cc / glibc directly defining `fpclassify` -> `__builtin_fpclassify`

**Comment by kroening** [comment]
We are eliminating the difference between goto-cc and cbmc's frontend!

File: src/ansi-c/c_typecheck_expr.cpp (hunk 7 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_expr.cpp]: done

File: src/ansi-c/c_typecheck_expr.cpp (hunk 8 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_expr.cpp]: done

**Comment by smowton** [comment]
Sure, what I mean is no test is currently exercising `__builting_fpclassify`, which you will get from e.g. this quote from glibc: https://github.com/bminor/glibc/blob/master/math/math.h#L885

**Review by martin-cs — APPROVED** [review]
I believe this to be technically correct, although I don't get the wider discussion of why it is needed.

**Comment by kroening** [comment]
Added a test

**Review by smowton — APPROVED** [review]
Thanks for adding the test!

**Pull request merged**